src/server/pfs/server/driver.go

```diff
@@ -939,191 +773,122 @@ func (d *driver) propagateCommits(sqlTx *sqlx.Tx, job *pfs.Job, branches []*pfs.
+		for _, provOfSubvB := range subvBI.Provenance {
+			provOfSubvBI, err := getBranchInfo(provOfSubvB)
+			if err != nil {
 				return err
 			}
-			for _, provProv := range provOfSubvBHeadInfo.Provenance {
-				newProvProv, _, err := d.resolveCommitProvenance(sqlTx, provProv)
-				if err != nil {
-					return errors.Wrapf(err, "could not resolve provenant commit %s@%s (%s)",
-						provProv.Commit.Branch.Repo.Name, provProv.Commit.Branch.Name, provProv.Commit.ID)
-				}
-				provProv = newProvProv
-				newCommitProvMap[commitKey(provProv.Commit)] = provProv
-				jobProvMap[pfsdb.CommitKey(provProv.Commit)] = provProv.Commit
-			}
-		}
-		if len(newCommitProvMap) == 0 {
-			// no input commits to process; don't create a new output commit
-			continue nextSubvBI
-		}
-
-		// 'subvB' may already have a HEAD commit, so compute whether the new output
-		// commit's provenance would be a subset of the existing HEAD commit's
-		// provenance. If so, a new output commit would be a duplicate, so don't
-		// create it.
-		if subvBI.Head != nil {
-			// get the info for subvB's HEAD commit
-			subvBHeadInfo := &pfs.CommitInfo{}
-			if err := d.commits.ReadWrite(sqlTx).Get(pfsdb.CommitKey(subvBI.Head), subvBHeadInfo); err != nil {
-				return pfsserver.ErrCommitNotFound{subvBI.Head}
-			}
-			provIntersection := make(map[string]struct{})
-			for _, p := range subvBHeadInfo.Provenance {
-				if _, ok := newCommitProvMap[commitKey(p.Commit)]; ok {
-					provIntersection[commitKey(p.Commit)] = struct{}{}
+			// If one of the provenances has a nil head, skip adding it to this Commitset
+			if provOfSubvBI.Head != nil {
+				if provOfSubvBI.Head.ID != txnCtx.CommitsetID {
+					if _, err := d.aliasCommit(txnCtx, provOfSubvBI.Head, provOfSubvBI.Head.Branch); err != nil {
+						return err
+					}
+					// Update the cached branch head
+					provOfSubvBI.Head.ID = txnCtx.CommitsetID
 				}
-			}
-			if len(newCommitProvMap) == len(provIntersection) {
-				// newCommit's provenance is subset of existing HEAD's provenance
-				continue nextSubvBI
+				// Generate the new Commitset commit info for this branch
+				addBranch(provOfSubvBI)
 			}
 		}
 
-		// If the only branches in the hypothetical output commit's provenance are
-		// in the 'spec' repo, creating it would mean creating a confusing "dummy"
-		// job with no non-spec input data. If this is the case, don't create a new
-		// output commit
-		allSpec := true
-		for _, p := range newCommitProvMap {
-			if p.Commit.Branch.Repo.Name != ppsconsts.SpecRepo {
-				allSpec = false
-				break
+		if subvBI.Head == nil || subvBI.Head.ID != txnCtx.CommitsetID {
+			// This branch has no commit for this Commitset, start a new output commit in 'subvBI.Branch'
+			newCommit := &pfs.Commit{
+				Branch: subvBI.Branch,
+				ID:     txnCtx.CommitsetID,
+			}
+			newCommitInfo := &pfs.CommitInfo{
+				Commit:  newCommit,
+				Origin:  &pfs.CommitOrigin{Kind: pfs.OriginKind_AUTO},
+				Started: types.TimestampNow(),
```

Comment:
I kind of wonder if it would make more sense for the timestamp to be the same for all of the commits in the transaction / commitset?

Reply:
That sounds reasonable to me, it's something we can set in the transaction context.  I'll do that.